react/LocationSelect.tsx

```diff
@@ -0,0 +1,118 @@
+import React from 'react'
+import { useAddressContext } from 'vtex.address-context/AddressContext'
+import rules from './countries/rules'
+import { Dropdown } from 'vtex.styleguide'
+import { FormattedMessage, defineMessages } from 'react-intl'
+
+const messages = defineMessages({
+  province: {
+    defaultMessage: '',
+    id: 'place-components.label.province',
+  },
+  city: {
+    defaultMessage: '',
+    id: 'place-components.label.city',
+  },
+  department: {
+    defaultMessage: '',
+    id: 'place-components.label.department',
+  },
+  region: {
+    defaultMessage: '',
+    id: 'place-components.label.region',
+  },
+  community: {
+    defaultMessage: '',
+    id: 'place-components.label.community',
+  },
+  municipality: {
+    defaultMessage: '',
+    id: 'place-components.label.municipality',
+  },
+  district: {
+    defaultMessage: '',
+    id: 'place-components.label.district',
+  },
+  state: {
+    defaultMessage: '',
+    id: 'place-components.label.state',
+  },
+  locality: {
+    defaultMessage: '',
+    id: 'place-components.label.locality',
+  },
+})
+
+const LocationSelect: StorefrontFunctionComponent<{}> = () => {
+  const { address, setAddress } = useAddressContext()
+  const countryRules = rules[address.country]
+
+  if (!countryRules.locationSelect)
+    throw 'The LocationSelect component is not applicable to this country :('
+
+  const { countryData, fields } = countryRules.locationSelect
+
+  const getCompletedFieldsArray = () => {
+    let completedFields = []
```

Comment:
This can be a `const`.

Reply:
I read that it's not a good idea to use `const` for non-primitive types (objects, arrays, functions), since one would expect that a `const` variable doesn't change its value, but the `const` keyword only adds the restriction that the variable can't be reassigned (so if it is an object/array/function, its value can still change).
What do you think about this? It seems that the pattern we use is using `const` when it's a variable we only use once.

Reply:
The argument that `const` could be misleading in that the variable can change makes sense, and I'd like to read about it if you still have the source, but I still think a `const` is better here. The reason is exactly what you said, it prevents the variable from being reassigned. Variable reassigning is problematic because sometimes it means you're using the same variable for different things, which can lead to bugs and a code that is harder to understand.

Reply:
I prefer `const` because it makes the code easier to reason about. Since it can't be reassigned,  you will have less side effects or edge cases to reason about.

Reply:
The source is [YDKJS](https://github.com/getify/You-Dont-Know-JS/blob/2nd-ed/get-started/ch2.md) (start reading from the first appearance of `const`).

I agree that avoiding reassignments is still good, and now that I'm thinking, maybe linters help with noticing mutations? 🤔 

Anyways, just to be extra clear with what kind of "undesirable" behavior you could get, here goes an example:

```javascript
const x = [1, 2, 3]
x[1] = 5
console.log(x)
```

Would show the array `[1, 5, 3]` while you would expect it to forbid you from making that change in the array.

Reply:
Well, someone aware of the true meaning of `const` would not expect JS to forbid the array mutation =P
But I guess `let` is fine in this situation. It is the most common keyword in this construct anyway.

Reply:
> someone aware of the true meaning of const would not expect JS to forbid the array mutation

That's exactly what I thought 😅 Well, this code won't have `let` anyways because it has been rewritten with a more functional style (as we discussed in the other conversation) that also helps in avoiding mutations.